manager/eris-mint/evm/snative.go

```diff
@@ -1,104 +1,351 @@
 package vm
 
 import (
-	"encoding/hex"
 	"fmt"
 
 	"github.com/eris-ltd/eris-db/common/sanity"
 	"github.com/eris-ltd/eris-db/manager/eris-mint/evm/sha3"
 	ptypes "github.com/eris-ltd/eris-db/permission/types"
 	. "github.com/eris-ltd/eris-db/word256"
+
+	"bytes"
+	"strings"
 )
 
 //------------------------------------------------------------------------------------------------
 // Registered SNative contracts
 
-var PermissionsContract = "permissions_contract"
-
-func registerSNativeContracts() {
-	registeredNativeContracts[LeftPadWord256([]byte(PermissionsContract))] = permissionsContract
-
-	/*
-		// we could expose these but we moved permission and args checks into the permissionsContract
-		// so calling them would be unsafe ...
-		registeredNativeContracts[LeftPadWord256([]byte("has_base"))] = has_base
-		registeredNativeContracts[LeftPadWord256([]byte("set_base"))] = set_base
-		registeredNativeContracts[LeftPadWord256([]byte("unset_base"))] = unset_base
-		registeredNativeContracts[LeftPadWord256([]byte("set_global"))] = set_global
-		registeredNativeContracts[LeftPadWord256([]byte("has_role"))] = has_role
-		registeredNativeContracts[LeftPadWord256([]byte("add_role"))] = add_role
-		registeredNativeContracts[LeftPadWord256([]byte("rm_role"))] = rm_role
-	*/
+type SNativeContractDescription struct {
+	Comment   string
+	Name      string
+	functions map[FuncID]SNativeFuncDescription
 }
 
-//-----------------------------------------------------------------------------
-// snative are native contracts that can access and modify an account's permissions
-
 type SNativeFuncDescription struct {
```

Comment:
Second thought / correction: `SNativeFuncDescription` should not become an interface as suggested above.  Rather the SNative definition should be defined with `LanguageTypes` and the templating should be a package on itself, that produces the solidity code based on the mapping of `LanguageType` to `SolidityType`.  We can raise this as an issue and proceed as is for now; the net gain in readability is sufficient for me.